added helper functions lesson (#37)
Ada-Developers-Academy/ada-precourse-v2: 01-intro-to-python/02.5-helper-functions.md

At the end of this lesson we will be able to:

- Create helper functions
- Refactor larger functions into smaller pieces

## Introduction

A helper function is a function that does part of the work for another function. They make our code easier to read by breaking up long expressions or functions into smaller pieces. We recommend giving helper functions descriptive names, to help with readability.

### !callout-info

## Functions Best Practice

> Ideally, every function should be designed to handle *one* task in accordance to the (single-responsibility principle)[https://en.wikipedia.org/wiki/Single-responsibility_principle]. Helper functions 

Reviewer: ```suggestion
Ideally, every function should be designed to handle *one* task in accordance to the (single-responsibility principle)[https://en.wikipedia.org/wiki/Single-responsibility_principle].  
```

> Ideally, every function should be designed to handle *one* task in accordance to the (single-responsibility principle)[https://en.wikipedia.org/wiki/Single-responsibility_principle]. Helper functions 

Reviewer 2: looks like there's a trailing sentence. 

```suggestion
Ideally, every function should be designed to handle *one* task in accordance to the (single-responsibility principle)[https://en.wikipedia.org/wiki/Single-responsibility_principle].
```

### !end-callout

## Breaking Up Long Functions

> Image we are building an ecommerce webapp that will display an order summary. In order to create this summary, two calculations must be performed: subtotal of all items purchased and sales tax. It would be reasonable to place these calculations into two separate helper functions `calculate_subtotal` and `calculate_sales_tax`. 

Reviewer 2: ```suggestion
Image we are building an ecommerce webapp that will display an order summary. In order to create this summary, two calculations must be performed: subtotal of all items purchased and sales tax. It would be reasonable to place these calculations into two separate helper functions: `calculate_subtotal` and `calculate_sales_tax`. 
```

    sub_total = calculate_subtotal(item_prices)
    sales_tax_total = calculate_sales_tax(sub_total) 
    grand_total = sub_total + sales_tax_total

    sub_total = "$ {:.2f}".format(sub_total)
    grand_total = "$ {:.2f}".format(grand_total)
    sales_tax_total = "$  {:.2f}".format(sales_tax_total)

    return (f"""
        **** Order Summary ****\n  
        Item(s) Subtotal:  {sub_total}
        Sales Tax:         {sales_tax_total}
        --------------
        Grand Total:       {grand_total}""") 
```

### !callout-info

## Assigning Variables to Function Calls

Let's recall our learnings from the `Return keyword` section of the previous Functions. If a function has a return value, we can assign a variable to its function call and expect a value. Let's breakdown the `sub_total` variable assignment in `display_order_summary`:

    - `sub_total` is assigned to the function call, `calculate_subtotal`.
    - `calculate_subtotal` runs and returns a value of 40 when finished.
    -  40 is now assigned to `subtotal`. 

    ```Python
    # before executing
    sub_total = calculate_subtotal(item_prices)
    # after executing 
    sub_total = 40 
    ```

> ```

Reviewer 2: I can't tell... is this an extra set of backticks?

```suggestion

```

### !end-callout

## Breaking Up Long Expressions 

Sometimes code will have long expressions that are not easy to read. Imagine we are writing a function to mimic the order of operations rule, (PEMDAS)[https://en.wikipedia.org/wiki/Order_of_operations#Mnemonics]. We could write the function as a long expression like so:

```Python
def pemdas(n1,n2,n3,n4,n5,n6):
    result = ((((n1**n2)*n3)/n4)+n5)-n6
    return result
```

> What happens if the exponent or the division is inaccurate within this expression? It would be mucher harder to test each calculation within this expression. To improve the on overall readability, what helper functions should we add? (Note: exclude detecting parenthesis)

Reviewer 2: ```suggestion
What happens if the exponent or the division is inaccurate within this expression? It would be much harder to test each calculation within this expression. To improve overall readability, what helper functions should we add? (Note: exclude detecting parenthesis)
```

<details>
<summary>click here to see the helper functions
```Python
    def multiply(n1, n2):
        return n1*n2

    def add(n1,n2):
        return n1+n2

    def subtract(n1,n2):
        return n1-n2

    def divide(n1,n2):
        return n1 / n2

    def exponent(n1, n2):
        return n1**n2 
    ```
</summary>
</details>

> ## Exercise:  Breaking Up `guess_the_number`

Reviewer: ```suggestion
## Guess The Number Project

Open up your `game.py` in VSCode and let's see how we can apply helper functions to our project!

### Exercise:  Breaking Up `guess_the_number`
```
